# Exploratory Data Analysis of IPL-Matches(2008-2022)
As you all know Cricket is a sport played and watched by millions of people around the world but Indians are mad about cricket. Cricket is played in our streets. We like it so much. People in india are crazy about the game. Cricket is a Gentlemen Game , As a cricket lover, I decided to analyze the data of IPL matches. Why IPL match? Because cricket is not a game of any specific country or a nation, I love Indian league IPL .So, that why I shall anlyze the data of IPL matches from 2008 to 2022.

I found IPL matches (2008-2022) dataset from kaggle.
- Here is the link of IPL matches Dataset: https://www.kaggle.com/datasets/vora1011/ipl-2008-to-2021-all-match-dataset?select=IPL_Matches_2008_2022.csv

## Data Selection and Installing 
First we will select appropriate dataset from kaggle which I have done. You can check it at this link: https://www.kaggle.com/datasets/vora1011/ipl-2008-to-2021-all-match-dataset?select=IPL_Matches_2008_2022.csv. To download our dataset we have to install opendataset library by pip install command.

In [2]:
dataset_url = 'https://www.kaggle.com/datasets/vora1011/ipl-2008-to-2021-all-match-dataset?select=IPL_Matches_2008_2022.csv'

## Essential Libraries for this Project
Lets import essential libraries for this project

In [3]:
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
#settings
sns.set_style('darkgrid')
matplotlib.rcParams['font.size'] = 14
matplotlib.rcParams['figure.figsize'] = (9, 5)
matplotlib.rcParams['figure.facecolor'] = '#00000000'

In [4]:
import warnings
warnings.filterwarnings("ignore")

## Analysing IPL match dataset

## Data Preparation and Cleaning
First, we load our dataset into DataFrame and see our DataFrame some first and last rows to overview of dataset

In [5]:

df = pd.read_csv('IPL_Matches_2008_2022.csv')


In [6]:
df.columns

Index(['ID', 'City', 'Date', 'Season', 'MatchNumber', 'Team1', 'Team2',
       'Venue', 'TossWinner', 'TossDecision', 'SuperOver', 'WinningTeam',
       'WonBy', 'Margin', 'method', 'Player_of_Match', 'Team1Players',
       'Team2Players', 'Umpire1', 'Umpire2'],
      dtype='object')

By looking at the dataset we can see that our dataset contains many rows and columns. The dataset includes **MatchNumber**, **Team1**, **Team2**, **Venue**, **Team1Players**, **Team2Players**, **Umpire1** and **Umpire2** Data which gives us information about where match held, which teams played a match and about team players, about umpires also and information about **winning team** and win stats.

In [7]:
df.shape

(950, 20)

There is 950 rows and 20 columns in our Dataset

#city , supperover , winningteam,margin,method,play of match

In [8]:
df.head(3)

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
0,1312200,Ahmedabad,2022-05-29,2022,Final,Rajasthan Royals,Gujarat Titans,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,bat,N,Gujarat Titans,Wickets,7.0,NaN,HH Pandya,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",CB Gaffaney,Nitin Menon
1,1312199,Ahmedabad,2022-05-27,2022,Qualifier 2,Royal Challengers Bangalore,Rajasthan Royals,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,field,N,Rajasthan Royals,Wickets,7.0,NaN,JC Buttler,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...",CB Gaffaney,Nitin Menon
2,1312198,Kolkata,2022-05-25,2022,Eliminator,Royal Challengers Bangalore,Lucknow Super Giants,"Eden Gardens, Kolkata",Lucknow Super Giants,field,N,Royal Challengers Bangalore,Runs,14.0,NaN,RM Patidar,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda...",J Madanagopal,MA Gough


In [9]:
df['WonBy'].nunique()

4

In [10]:
df.value_counts()

ID       City           Date        Season   MatchNumber  Team1                        Team2                        Venue                                                TossWinner                   TossDecision  SuperOver  WinningTeam                  WonBy    Margin  method  Player_of_Match   Team1Players                                                                                                                                               Team2Players                                                                                                                                                 Umpire1                Umpire2        
336022   Delhi          2008-05-17  2007/08  40           Delhi Daredevils             Kings XI Punjab              Feroz Shah Kotla                                     Delhi Daredevils             bat           N          Kings XI Punjab              Runs     6.0     D/L     DPMD Jayawardene  ['G Gambhir', 'V Sehwag', 'S Dhawan', 'TM Dilshan', 'MF Mah

In [11]:
df.describe()['Margin']

count    932.000000
mean      17.056867
std       21.633109
min        1.000000
25%        6.000000
50%        8.000000
75%       19.000000
max      146.000000
Name: Margin, dtype: float64

In [12]:
# groupby
df.groupby('WonBy')['Margin'].mean()

WonBy
NoResults          NaN
Runs         30.073286
SuperOver          NaN
Wickets       6.239686
Name: Margin, dtype: float64

In [13]:
# groupby
df.groupby('WonBy')['Margin'].median()

WonBy
NoResults     NaN
Runs         22.0
SuperOver     NaN
Wickets       6.0
Name: Margin, dtype: float64

### I cant se about mean,median and other statistical value > bcoZ depended on wonby col(runs,wickets,superover,noresult)

In [14]:
df.nunique()

ID                 950
City                33
Date               706
Season              15
MatchNumber         80
Team1               18
Team2               18
Venue               49
TossWinner          18
TossDecision         2
SuperOver            2
WinningTeam         18
WonBy                4
Margin              94
method               1
Player_of_Match    262
Team1Players       932
Team2Players       937
Umpire1             58
Umpire2             60
dtype: int64

#### The above table provides us with very useful information. By above table we can see that as of 2022, there have been 15 seasons of IPL tournament hosted by 33 cities. Although, 18 teams have played and 262 Players become player of match in past seasons of IPL.

In [15]:
df.isnull().sum()

ID                   0
City                51
Date                 0
Season               0
MatchNumber          0
Team1                0
Team2                0
Venue                0
TossWinner           0
TossDecision         0
SuperOver            4
WinningTeam          4
WonBy                0
Margin              18
method             931
Player_of_Match      4
Team1Players         0
Team2Players         0
Umpire1              0
Umpire2              0
dtype: int64

In [16]:
print("Total Null Cells",df.isnull().sum().sum())
df.isnull().sum()

Total Null Cells 1012


ID                   0
City                51
Date                 0
Season               0
MatchNumber          0
Team1                0
Team2                0
Venue                0
TossWinner           0
TossDecision         0
SuperOver            4
WinningTeam          4
WonBy                0
Margin              18
method             931
Player_of_Match      4
Team1Players         0
Team2Players         0
Umpire1              0
Umpire2              0
dtype: int64

In [17]:
df.Date.dtype #

dtype('O')

In [18]:
df.columns

Index(['ID', 'City', 'Date', 'Season', 'MatchNumber', 'Team1', 'Team2',
       'Venue', 'TossWinner', 'TossDecision', 'SuperOver', 'WinningTeam',
       'WonBy', 'Margin', 'method', 'Player_of_Match', 'Team1Players',
       'Team2Players', 'Umpire1', 'Umpire2'],
      dtype='object')

In [19]:
df.Date 

0      2022-05-29
1      2022-05-27
2      2022-05-25
3      2022-05-24
4      2022-05-22
          ...    
945    2008-04-20
946    2008-04-20
947    2008-04-19
948    2008-04-19
949    2008-04-18
Name: Date, Length: 950, dtype: object

In our DataFrame the date should be a string that represent date for future work with date. Let's convert all cells in **Date** column to dates. Pandas have **datetime()** method for this. 

In [20]:
pd.to_datetime(df['Date'])

0     2022-05-29
1     2022-05-27
2     2022-05-25
3     2022-05-24
4     2022-05-22
         ...    
945   2008-04-20
946   2008-04-20
947   2008-04-19
948   2008-04-19
949   2008-04-18
Name: Date, Length: 950, dtype: datetime64[ns]

In [21]:
df['Date']

0      2022-05-29
1      2022-05-27
2      2022-05-25
3      2022-05-24
4      2022-05-22
          ...    
945    2008-04-20
946    2008-04-20
947    2008-04-19
948    2008-04-19
949    2008-04-18
Name: Date, Length: 950, dtype: object

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 950 entries, 0 to 949
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               950 non-null    int64  
 1   City             899 non-null    object 
 2   Date             950 non-null    object 
 3   Season           950 non-null    object 
 4   MatchNumber      950 non-null    object 
 5   Team1            950 non-null    object 
 6   Team2            950 non-null    object 
 7   Venue            950 non-null    object 
 8   TossWinner       950 non-null    object 
 9   TossDecision     950 non-null    object 
 10  SuperOver        946 non-null    object 
 11  WinningTeam      946 non-null    object 
 12  WonBy            950 non-null    object 
 13  Margin           932 non-null    float64
 14  method           19 non-null     object 
 15  Player_of_Match  946 non-null    object 
 16  Team1Players     950 non-null    object 
 17  Team2Players    

In [23]:
df['Date'][0]

'2022-05-29'

In [24]:
df['Date']

0      2022-05-29
1      2022-05-27
2      2022-05-25
3      2022-05-24
4      2022-05-22
          ...    
945    2008-04-20
946    2008-04-20
947    2008-04-19
948    2008-04-19
949    2008-04-18
Name: Date, Length: 950, dtype: object

Now, we will clean empty cells, wrong data and unusual data. We will also check any duplicate data in our dataset.

In [25]:
### ceking duplicate rows
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
945    False
946    False
947    False
948    False
949    False
Length: 950, dtype: bool

In [26]:
### ceking duplicate rows
df.duplicated().sum()

np.int64(0)

> Thera are no duplicate within the dateset

In [27]:
df.columns

Index(['ID', 'City', 'Date', 'Season', 'MatchNumber', 'Team1', 'Team2',
       'Venue', 'TossWinner', 'TossDecision', 'SuperOver', 'WinningTeam',
       'WonBy', 'Margin', 'method', 'Player_of_Match', 'Team1Players',
       'Team2Players', 'Umpire1', 'Umpire2'],
      dtype='object')

In [28]:
df.value_counts()


ID       City           Date        Season   MatchNumber  Team1                        Team2                        Venue                                                TossWinner                   TossDecision  SuperOver  WinningTeam                  WonBy    Margin  method  Player_of_Match   Team1Players                                                                                                                                               Team2Players                                                                                                                                                 Umpire1                Umpire2        
336022   Delhi          2008-05-17  2007/08  40           Delhi Daredevils             Kings XI Punjab              Feroz Shah Kotla                                     Delhi Daredevils             bat           N          Kings XI Punjab              Runs     6.0     D/L     DPMD Jayawardene  ['G Gambhir', 'V Sehwag', 'S Dhawan', 'TM Dilshan', 'MF Mah

In [29]:
df['City'].value_counts()


City
Mumbai            159
Kolkata            79
Delhi              78
Chennai            67
Bangalore          65
Hyderabad          64
Chandigarh         56
Pune               51
Jaipur             47
Abu Dhabi          37
Ahmedabad          19
Bengaluru          15
Durban             15
Visakhapatnam      13
Dubai              13
Centurion          12
Sharjah            10
Rajkot             10
Navi Mumbai         9
Indore              9
Dharamsala          9
Johannesburg        8
Port Elizabeth      7
Cuttack             7
Ranchi              7
Cape Town           7
Raipur              6
Kochi               5
Kanpur              4
Nagpur              3
Kimberley           3
East London         3
Bloemfontein        2
Name: count, dtype: int64

In [30]:
len(df['City'].value_counts())


33

In [31]:
df['City'].value_counts().head(33)

City
Mumbai            159
Kolkata            79
Delhi              78
Chennai            67
Bangalore          65
Hyderabad          64
Chandigarh         56
Pune               51
Jaipur             47
Abu Dhabi          37
Ahmedabad          19
Bengaluru          15
Durban             15
Visakhapatnam      13
Dubai              13
Centurion          12
Sharjah            10
Rajkot             10
Navi Mumbai         9
Indore              9
Dharamsala          9
Johannesburg        8
Port Elizabeth      7
Cuttack             7
Ranchi              7
Cape Town           7
Raipur              6
Kochi               5
Kanpur              4
Nagpur              3
Kimberley           3
East London         3
Bloemfontein        2
Name: count, dtype: int64

In [32]:
for City in df['City'].value_counts().index:
    print(City)


Mumbai
Kolkata
Delhi
Chennai
Bangalore
Hyderabad
Chandigarh
Pune
Jaipur
Abu Dhabi
Ahmedabad
Bengaluru
Durban
Visakhapatnam
Dubai
Centurion
Sharjah
Rajkot
Navi Mumbai
Indore
Dharamsala
Johannesburg
Port Elizabeth
Cuttack
Ranchi
Cape Town
Raipur
Kochi
Kanpur
Nagpur
Kimberley
East London
Bloemfontein


In [33]:
df['City'].value_counts().sort_index()

City
Abu Dhabi          37
Ahmedabad          19
Bangalore          65
Bengaluru          15
Bloemfontein        2
Cape Town           7
Centurion          12
Chandigarh         56
Chennai            67
Cuttack             7
Delhi              78
Dharamsala          9
Dubai              13
Durban             15
East London         3
Hyderabad          64
Indore              9
Jaipur             47
Johannesburg        8
Kanpur              4
Kimberley           3
Kochi               5
Kolkata            79
Mumbai            159
Nagpur              3
Navi Mumbai         9
Port Elizabeth      7
Pune               51
Raipur              6
Rajkot             10
Ranchi              7
Sharjah            10
Visakhapatnam      13
Name: count, dtype: int64

In [34]:
df['City'].replace

<bound method NDFrame.replace of 0       Ahmedabad
1       Ahmedabad
2         Kolkata
3         Kolkata
4          Mumbai
          ...    
945       Kolkata
946        Mumbai
947         Delhi
948    Chandigarh
949     Bangalore
Name: City, Length: 950, dtype: object>

In [35]:
f1 = df['City']=='Bangalore'

In [36]:
df[f1].replace('')

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
315,1082649,Bangalore,2017-05-19,2017,Qualifier 2,Mumbai Indians,Kolkata Knight Riders,M Chinnaswamy Stadium,Mumbai Indians,field,N,Mumbai Indians,Wickets,6.0,NaN,KV Sharma,"['LMP Simmons', 'PA Patel', 'AT Rayudu', 'RG S...","['CA Lynn', 'SP Narine', 'G Gambhir', 'RV Utha...",NJ Llong,Nitin Menon
316,1082648,Bangalore,2017-05-17,2017,Eliminator,Sunrisers Hyderabad,Kolkata Knight Riders,M Chinnaswamy Stadium,Kolkata Knight Riders,field,N,Kolkata Knight Riders,Wickets,7.0,D/L,NM Coulter-Nile,"['DA Warner', 'S Dhawan', 'KS Williamson', 'Yu...","['RV Uthappa', 'CA Lynn', 'YK Pathan', 'G Gamb...",AK Chaudhary,Nitin Menon
327,1082636,Bangalore,2017-05-07,2017,46,Royal Challengers Bangalore,Kolkata Knight Riders,M Chinnaswamy Stadium,Kolkata Knight Riders,field,N,Kolkata Knight Riders,Wickets,6.0,NaN,SP Narine,"['CH Gayle', 'Mandeep Singh', 'V Kohli', 'AB d...","['CA Lynn', 'SP Narine', 'C de Grandhomme', 'G...",AY Dandekar,C Shamshuddin
331,1082633,Bangalore,2017-05-05,2017,43,Royal Challengers Bangalore,Kings XI Punjab,M Chinnaswamy Stadium,Royal Challengers Bangalore,field,N,Kings XI Punjab,Runs,19.0,NaN,Sandeep Sharma,"['Mandeep Singh', 'CH Gayle', 'V Kohli', 'AB d...","['HM Amla', 'MJ Guptill', 'SE Marsh', 'M Vohra...",CB Gaffaney,C Shamshuddin
343,1082621,Bangalore,2017-04-27,2017,31,Royal Challengers Bangalore,Gujarat Lions,M Chinnaswamy Stadium,Gujarat Lions,field,N,Gujarat Lions,Wickets,7.0,NaN,AJ Tye,"['CH Gayle', 'V Kohli', 'AB de Villiers', 'TM ...","['Ishan Kishan', 'BB McCullum', 'SK Raina', 'A...",AK Chaudhary,C Shamshuddin
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
925,336006,Bangalore,2008-05-05,2007/08,25,Royal Challengers Bangalore,Kings XI Punjab,M Chinnaswamy Stadium,Kings XI Punjab,field,N,Kings XI Punjab,Wickets,6.0,NaN,S Sreesanth,"['W Jaffer', 'V Kohli', 'CL White', 'JH Kallis...","['SE Marsh', 'JR Hopes', 'RR Sarwan', 'Yuvraj ...",SJ Davis,BR Doctrove
928,336034,Bangalore,2008-05-03,2007/08,21,Royal Challengers Bangalore,Deccan Chargers,M Chinnaswamy Stadium,Deccan Chargers,field,N,Royal Challengers Bangalore,Runs,3.0,NaN,P Kumar,"['B Chipli', 'W Jaffer', 'V Kohli', 'JH Kallis...","['AC Gilchrist', 'VVS Laxman', 'HH Gibbs', 'RG...",BR Doctrove,SL Shastri
935,335996,Bangalore,2008-04-28,2007/08,15,Royal Challengers Bangalore,Chennai Super Kings,M Chinnaswamy Stadium,Chennai Super Kings,bat,N,Chennai Super Kings,Runs,13.0,NaN,MS Dhoni,"['B Chipli', 'W Jaffer', 'LRPL Taylor', 'JH Ka...","['PA Patel', 'ML Hayden', 'MEK Hussey', 'SK Ra...",BR Doctrove,RB Tiffin
939,335992,Bangalore,2008-04-26,2007/08,12,Royal Challengers Bangalore,Rajasthan Royals,M Chinnaswamy Stadium,Rajasthan Royals,field,N,Rajasthan Royals,Wickets,7.0,NaN,SR Watson,"['S Chanderpaul', 'R Dravid', 'LRPL Taylor', '...","['GC Smith', 'M Rawat', 'YK Pathan', 'SR Watso...",MR Benson,IL Howell


In [37]:
df[f1].replace('Bangalore','Bangalore')

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
315,1082649,Bangalore,2017-05-19,2017,Qualifier 2,Mumbai Indians,Kolkata Knight Riders,M Chinnaswamy Stadium,Mumbai Indians,field,N,Mumbai Indians,Wickets,6.0,NaN,KV Sharma,"['LMP Simmons', 'PA Patel', 'AT Rayudu', 'RG S...","['CA Lynn', 'SP Narine', 'G Gambhir', 'RV Utha...",NJ Llong,Nitin Menon
316,1082648,Bangalore,2017-05-17,2017,Eliminator,Sunrisers Hyderabad,Kolkata Knight Riders,M Chinnaswamy Stadium,Kolkata Knight Riders,field,N,Kolkata Knight Riders,Wickets,7.0,D/L,NM Coulter-Nile,"['DA Warner', 'S Dhawan', 'KS Williamson', 'Yu...","['RV Uthappa', 'CA Lynn', 'YK Pathan', 'G Gamb...",AK Chaudhary,Nitin Menon
327,1082636,Bangalore,2017-05-07,2017,46,Royal Challengers Bangalore,Kolkata Knight Riders,M Chinnaswamy Stadium,Kolkata Knight Riders,field,N,Kolkata Knight Riders,Wickets,6.0,NaN,SP Narine,"['CH Gayle', 'Mandeep Singh', 'V Kohli', 'AB d...","['CA Lynn', 'SP Narine', 'C de Grandhomme', 'G...",AY Dandekar,C Shamshuddin
331,1082633,Bangalore,2017-05-05,2017,43,Royal Challengers Bangalore,Kings XI Punjab,M Chinnaswamy Stadium,Royal Challengers Bangalore,field,N,Kings XI Punjab,Runs,19.0,NaN,Sandeep Sharma,"['Mandeep Singh', 'CH Gayle', 'V Kohli', 'AB d...","['HM Amla', 'MJ Guptill', 'SE Marsh', 'M Vohra...",CB Gaffaney,C Shamshuddin
343,1082621,Bangalore,2017-04-27,2017,31,Royal Challengers Bangalore,Gujarat Lions,M Chinnaswamy Stadium,Gujarat Lions,field,N,Gujarat Lions,Wickets,7.0,NaN,AJ Tye,"['CH Gayle', 'V Kohli', 'AB de Villiers', 'TM ...","['Ishan Kishan', 'BB McCullum', 'SK Raina', 'A...",AK Chaudhary,C Shamshuddin
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
925,336006,Bangalore,2008-05-05,2007/08,25,Royal Challengers Bangalore,Kings XI Punjab,M Chinnaswamy Stadium,Kings XI Punjab,field,N,Kings XI Punjab,Wickets,6.0,NaN,S Sreesanth,"['W Jaffer', 'V Kohli', 'CL White', 'JH Kallis...","['SE Marsh', 'JR Hopes', 'RR Sarwan', 'Yuvraj ...",SJ Davis,BR Doctrove
928,336034,Bangalore,2008-05-03,2007/08,21,Royal Challengers Bangalore,Deccan Chargers,M Chinnaswamy Stadium,Deccan Chargers,field,N,Royal Challengers Bangalore,Runs,3.0,NaN,P Kumar,"['B Chipli', 'W Jaffer', 'V Kohli', 'JH Kallis...","['AC Gilchrist', 'VVS Laxman', 'HH Gibbs', 'RG...",BR Doctrove,SL Shastri
935,335996,Bangalore,2008-04-28,2007/08,15,Royal Challengers Bangalore,Chennai Super Kings,M Chinnaswamy Stadium,Chennai Super Kings,bat,N,Chennai Super Kings,Runs,13.0,NaN,MS Dhoni,"['B Chipli', 'W Jaffer', 'LRPL Taylor', 'JH Ka...","['PA Patel', 'ML Hayden', 'MEK Hussey', 'SK Ra...",BR Doctrove,RB Tiffin
939,335992,Bangalore,2008-04-26,2007/08,12,Royal Challengers Bangalore,Rajasthan Royals,M Chinnaswamy Stadium,Rajasthan Royals,field,N,Rajasthan Royals,Wickets,7.0,NaN,SR Watson,"['S Chanderpaul', 'R Dravid', 'LRPL Taylor', '...","['GC Smith', 'M Rawat', 'YK Pathan', 'SR Watso...",MR Benson,IL Howell


Reenaming Team

In [38]:
df['Team1'].value_counts()

Team1
Royal Challengers Bangalore    126
Mumbai Indians                 112
Chennai Super Kings            111
Kolkata Knight Riders          105
Kings XI Punjab                 92
Rajasthan Royals                89
Delhi Daredevils                85
Sunrisers Hyderabad             69
Deccan Chargers                 39
Delhi Capitals                  29
Pune Warriors                   23
Punjab Kings                    18
Gujarat Lions                   16
Lucknow Super Giants             8
Gujarat Titans                   7
Rising Pune Supergiant           7
Rising Pune Supergiants          7
Kochi Tuskers Kerala             7
Name: count, dtype: int64

In [39]:
df['Team1']=df['Team1'].replace('Kings XI Punjab','Punjab Kings')

In [40]:
df['Team2']=df['Team2'].replace('Kings XI Punjab','Punjab Kings')

In [41]:
df['Team1'].value_counts()

Team1
Royal Challengers Bangalore    126
Mumbai Indians                 112
Chennai Super Kings            111
Punjab Kings                   110
Kolkata Knight Riders          105
Rajasthan Royals                89
Delhi Daredevils                85
Sunrisers Hyderabad             69
Deccan Chargers                 39
Delhi Capitals                  29
Pune Warriors                   23
Gujarat Lions                   16
Lucknow Super Giants             8
Gujarat Titans                   7
Rising Pune Supergiant           7
Rising Pune Supergiants          7
Kochi Tuskers Kerala             7
Name: count, dtype: int64

In [42]:
df['WinningTeam']=df['WinningTeam'].replace('Kings XI Punjab','Punjab Kings')

In [43]:
df['WinningTeam'].value_counts()

WinningTeam
Mumbai Indians                 131
Chennai Super Kings            121
Kolkata Knight Riders          114
Royal Challengers Bangalore    109
Punjab Kings                   101
Rajasthan Royals                96
Sunrisers Hyderabad             75
Delhi Daredevils                67
Delhi Capitals                  36
Deccan Chargers                 29
Gujarat Lions                   13
Gujarat Titans                  12
Pune Warriors                   12
Rising Pune Supergiant          10
Lucknow Super Giants             9
Kochi Tuskers Kerala             6
Rising Pune Supergiants          5
Name: count, dtype: int64

In [44]:
df.columns

Index(['ID', 'City', 'Date', 'Season', 'MatchNumber', 'Team1', 'Team2',
       'Venue', 'TossWinner', 'TossDecision', 'SuperOver', 'WinningTeam',
       'WonBy', 'Margin', 'method', 'Player_of_Match', 'Team1Players',
       'Team2Players', 'Umpire1', 'Umpire2'],
      dtype='object')

In [45]:
df['TossWinner']=df['TossWinner'].replace('Kings XI Punjab','Punjab Kings')

In [46]:
df.TossWinner.value_counts()

TossWinner
Mumbai Indians                 123
Kolkata Knight Riders          114
Chennai Super Kings            109
Royal Challengers Bangalore    105
Rajasthan Royals                99
Punjab Kings                    94
Delhi Daredevils                80
Sunrisers Hyderabad             74
Deccan Chargers                 43
Delhi Capitals                  36
Pune Warriors                   20
Gujarat Lions                   15
Gujarat Titans                  10
Kochi Tuskers Kerala             8
Lucknow Super Giants             7
Rising Pune Supergiants          7
Rising Pune Supergiant           6
Name: count, dtype: int64

In [47]:
df['Team1']=df['Team1'].replace('Delhi Darredevils','Delhi capital')
df['Team2']=df['Team2'].replace('Delhi Darredevils','Delhi capital')
df['WinningTeam']=df['WinningTeam'].replace('Delhi Darredevils','Delhi capital')

In [48]:
df['Team1']=df['Team1'].replace('Rising Pune Supergiants','Rising Pune Supergiant')
df['Team2']=df['Team2'].replace('Rising Pune Supergiants','Rising Pune Supergiant')
df['WinningTeam']=df['WinningTeam'].replace('Rising Pune Supergiants','Rising Pune Supergiant')

Dealing with null value

In [49]:
df.isnull().sum()

ID                   0
City                51
Date                 0
Season               0
MatchNumber          0
Team1                0
Team2                0
Venue                0
TossWinner           0
TossDecision         0
SuperOver            4
WinningTeam          4
WonBy                0
Margin              18
method             931
Player_of_Match      4
Team1Players         0
Team2Players         0
Umpire1              0
Umpire2              0
dtype: int64

In [50]:
# city 
df['City'].isnull()

0      False
1      False
2      False
3      False
4      False
       ...  
945    False
946    False
947    False
948    False
949    False
Name: City, Length: 950, dtype: bool

In [51]:
C1 = df['City'].isnull()
df[C1].nunique()

ID                 51
City                0
Date               45
Season              2
MatchNumber        42
Team1               9
Team2               9
Venue               2
TossWinner          8
TossDecision        2
SuperOver           2
WinningTeam         9
WonBy               3
Margin             26
method              0
Player_of_Match    38
Team1Players       51
Team2Players       49
Umpire1            13
Umpire2            10
dtype: int64

In [52]:
C1 = df['City'].isnull()
df[C1]['Venue'].value_counts()

Venue
Dubai International Cricket Stadium    33
Sharjah Cricket Stadium                18
Name: count, dtype: int64

In [53]:
vf =(df['Venue']=='Dubai International Cricket Stadium')
df[vf]['City']

74     Dubai
77     Dubai
78     Dubai
81     Dubai
84     Dubai
86     Dubai
89     Dubai
91     Dubai
94     Dubai
96     Dubai
101    Dubai
102    Dubai
104    Dubai
134      NaN
137      NaN
140      NaN
142      NaN
145      NaN
147      NaN
150      NaN
151      NaN
154      NaN
156      NaN
159      NaN
160      NaN
164      NaN
165      NaN
167      NaN
170      NaN
172      NaN
175      NaN
176      NaN
180      NaN
182      NaN
184      NaN
187      NaN
188      NaN
191      NaN
192      NaN
532      NaN
534      NaN
539      NaN
540      NaN
542      NaN
546      NaN
547      NaN
Name: City, dtype: object

In [54]:
vf =(df['Venue']=='Dubai International Cricket Stadium')

df[vf]['City']='Dubai' 

In [55]:
df['City'].replace('')

0       Ahmedabad
1       Ahmedabad
2         Kolkata
3         Kolkata
4          Mumbai
          ...    
945       Kolkata
946        Mumbai
947         Delhi
948    Chandigarh
949     Bangalore
Name: City, Length: 950, dtype: object

In [56]:
# city 
c1=df['City'].isnull()
df[c1]['Venue'].value_counts()
#-----

Venue
Dubai International Cricket Stadium    33
Sharjah Cricket Stadium                18
Name: count, dtype: int64

In [57]:
vf1 = (df['Venue'] == 'Sharjah Cricket Stadium')
df.loc[vf1,'City'] = 'Sharjah'

In [58]:
df.isnull().sum()

ID                   0
City                33
Date                 0
Season               0
MatchNumber          0
Team1                0
Team2                0
Venue                0
TossWinner           0
TossDecision         0
SuperOver            4
WinningTeam          4
WonBy                0
Margin              18
method             931
Player_of_Match      4
Team1Players         0
Team2Players         0
Umpire1              0
Umpire2              0
dtype: int64

In [59]:
df['SuperOver'].fillna('NoResult',inplace=True)
df['WinningTeam'].fillna('NoResult',inplace=True)
df['Player_of_Match'].fillna('NoResult',inplace= True)
df['Margin'].fillna('NoMargin',inplace = True)

In [60]:
df.isnull().sum()

ID                   0
City                33
Date                 0
Season               0
MatchNumber          0
Team1                0
Team2                0
Venue                0
TossWinner           0
TossDecision         0
SuperOver            0
WinningTeam          0
WonBy                0
Margin               0
method             931
Player_of_Match      0
Team1Players         0
Team2Players         0
Umpire1              0
Umpire2              0
dtype: int64

In [61]:
df.drop(columns=['method'],inplace=True)

Date has been cleaned for further processing

In [64]:
df.coloumns

AttributeError: 'DataFrame' object has no attribute 'coloumns'

In [66]:
df.head()

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
0,1312200,Ahmedabad,2022-05-29,2022,Final,Rajasthan Royals,Gujarat Titans,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,bat,N,Gujarat Titans,Wickets,7.0,HH Pandya,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",CB Gaffaney,Nitin Menon
1,1312199,Ahmedabad,2022-05-27,2022,Qualifier 2,Royal Challengers Bangalore,Rajasthan Royals,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,field,N,Rajasthan Royals,Wickets,7.0,JC Buttler,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...",CB Gaffaney,Nitin Menon
2,1312198,Kolkata,2022-05-25,2022,Eliminator,Royal Challengers Bangalore,Lucknow Super Giants,"Eden Gardens, Kolkata",Lucknow Super Giants,field,N,Royal Challengers Bangalore,Runs,14.0,RM Patidar,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda...",J Madanagopal,MA Gough
3,1312197,Kolkata,2022-05-24,2022,Qualifier 1,Rajasthan Royals,Gujarat Titans,"Eden Gardens, Kolkata",Gujarat Titans,field,N,Gujarat Titans,Wickets,7.0,DA Miller,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",BNJ Oxenford,VK Sharma
4,1304116,Mumbai,2022-05-22,2022,70,Sunrisers Hyderabad,Punjab Kings,"Wankhede Stadium, Mumbai",Sunrisers Hyderabad,bat,N,Punjab Kings,Wickets,5.0,Harpreet Brar,"['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...","['JM Bairstow', 'S Dhawan', 'M Shahrukh Khan',...",AK Chaudhary,NA Patwardhan


In [67]:
dfb=df

In [68]:

dfb['Innings'].value_counts()

KeyError: 'Innings'

In [69]:
dfb.head()

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
0,1312200,Ahmedabad,2022-05-29,2022,Final,Rajasthan Royals,Gujarat Titans,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,bat,N,Gujarat Titans,Wickets,7.0,HH Pandya,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",CB Gaffaney,Nitin Menon
1,1312199,Ahmedabad,2022-05-27,2022,Qualifier 2,Royal Challengers Bangalore,Rajasthan Royals,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,field,N,Rajasthan Royals,Wickets,7.0,JC Buttler,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...",CB Gaffaney,Nitin Menon
2,1312198,Kolkata,2022-05-25,2022,Eliminator,Royal Challengers Bangalore,Lucknow Super Giants,"Eden Gardens, Kolkata",Lucknow Super Giants,field,N,Royal Challengers Bangalore,Runs,14.0,RM Patidar,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda...",J Madanagopal,MA Gough
3,1312197,Kolkata,2022-05-24,2022,Qualifier 1,Rajasthan Royals,Gujarat Titans,"Eden Gardens, Kolkata",Gujarat Titans,field,N,Gujarat Titans,Wickets,7.0,DA Miller,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",BNJ Oxenford,VK Sharma
4,1304116,Mumbai,2022-05-22,2022,70,Sunrisers Hyderabad,Punjab Kings,"Wankhede Stadium, Mumbai",Sunrisers Hyderabad,bat,N,Punjab Kings,Wickets,5.0,Harpreet Brar,"['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...","['JM Bairstow', 'S Dhawan', 'M Shahrukh Khan',...",AK Chaudhary,NA Patwardhan


In [70]:
dfb.value_counts()

ID       City        Date        Season   MatchNumber  Team1                        Team2                        Venue                                       TossWinner                   TossDecision  SuperOver  WinningTeam                  WonBy    Margin  Player_of_Match  Team1Players                                                                                                                                                Team2Players                                                                                                                                                  Umpire1      Umpire2       
1312200  Ahmedabad   2022-05-29  2022     Final        Rajasthan Royals             Gujarat Titans               Narendra Modi Stadium, Ahmedabad            Rajasthan Royals             bat           N          Gujarat Titans               Wickets  7.0     HH Pandya        ['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D Padikkal', 'SO Hetmyer', 'R Ashwin', 'R Parag', 'TA Boult', 'OC 

In [71]:
dfb.nunique()

ID                 950
City                33
Date               706
Season              15
MatchNumber         80
Team1               16
Team2               16
Venue               49
TossWinner          17
TossDecision         2
SuperOver            3
WinningTeam         17
WonBy                4
Margin              95
Player_of_Match    263
Team1Players       932
Team2Players       937
Umpire1             58
Umpire2             60
dtype: int64

In [79]:
dfb.columns

Index(['ID', 'City', 'Date', 'Season', 'MatchNumber', 'Team1', 'Team2',
       'Venue', 'TossWinner', 'TossDecision', 'SuperOver', 'WinningTeam',
       'WonBy', 'Margin', 'Player_of_Match', 'Team1Players', 'Team2Players',
       'Umpire1', 'Umpire2'],
      dtype='object')

In [83]:
dfb['BattingTeam'].value_counts().sort_index()

KeyError: 'BattingTeam'